File: Client/Chuangshi/Skill/Effect/HoldEffect/HoldEffectPlayEffect.cpp
#include "HoldEffectPlayEffect.h"
#include "Util/CS3Debug.h"
#include "Util/ConvertUtil.h"
#include "Util/GolbalBPFunctionLibrary.h"
#include "Interface/Monster/CSkillInterface.h"
#include "Interface/SkillInterface.h"
#include "CS3Base/GameObject.h"

UHoldEffectPlayEffect::UHoldEffectPlayEffect()
{
}

UHoldEffectPlayEffect::~UHoldEffectPlayEffect()
{
}

void UHoldEffectPlayEffect::init(FHOLD_EFFECT dictDat, USkill * skill)
{
	UHoldEffect::init(dictDat, skill);
	if (dictDat.Param1 != "")
	{
		TArray<FString> Param1 = UGolbalBPFunctionLibrary::SplitString(dictDat.Param1, false, TEXT("|"));
		for (auto i : Param1)
		{
			TArray<FString> actionInfo = UGolbalBPFunctionLibrary::SplitString(i, false, TEXT(":"));
			if (actionInfo.Num() > 1)
			{
				_actionID.Add(actionInfo[0]);
				_actionBeginTime.Add(actionInfo[1]);

			} else {
				_actionID.Add(actionInfo[0]);
				_actionBeginTime.Add("0.0");
			}
		}
	}
	_effectID = dictDat.Param2;
	_soundID = dictDat.Param3;
}

void UHoldEffectPlayEffect::onBegin(UHoldEffectDataType * effectData, CS3Entity * receiver)
{
	UHoldEffect::onBegin(effectData, receiver);
	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), receiver->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->setActionIndex();
		int32 ActionIndex = CSkillInterface->getActionIndex();
		effectData->setTempData("actionIndex", FVariant(ActionIndex));
		USkillInterface *SkillInterface = ((GameObject*)receiver)->Interface_Cache.SkillInterface;
		if (IsValid(SkillInterface))
		{
			SkillInterface->HoldEffectPlayEffect(effectData->casterID, effectData->index, ActionIndex, _actionID, _effectID, _soundID, _actionBeginTime);
		}
		CSkillInterface->BroadcastHoldEffectPlayEffect(effectData->casterID, effectData->index, ActionIndex, _actionID, _effectID, _soundID, _actionBeginTime);
	}
}

void UHoldEffectPlayEffect::onEnd(UHoldEffectDataType * effectData, CS3Entity * receiver)
{
	FVariant value = effectData->getTempData("actionIndex", FVariant(0));
	int32 ActionIndex = value.GetValue<int32>();
	USkillInterface *SkillInterface = ((GameObject*)receiver)->Interface_Cache.SkillInterface;
	if (IsValid(SkillInterface))
	{
		SkillInterface->HoldEffectEndEffect( effectData->index, ActionIndex);
	}
	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), receiver->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->BroadcastHoldEffectEndEffect(effectData->index, ActionIndex);
	}
	UHoldEffect::onEnd(effectData, receiver);
}

File: Server/scripts/cell/ConfigObject/Skill/Effect/HoldEffect/HoldEffectPlayEffect.py
# -*- coding: utf-8 -*-

import KBEDebug
import csdefine
import KBEngine
from ConfigObject.Skill.Effect.HoldEffect.HoldEffect import HoldEffect

class HoldEffectPlayEffect( HoldEffect ):
	"""
	播放动作、光效、音效
	"""
	def __init__( self ):
		HoldEffect.__init__( self )
		self._actionID = []
		self._effectID = ""
		self._soundID = ""
		self._actionBeginTime = []
	
	def init( self, dictDat, skill ):
		"""
		virtual method;
		读取技能配置
		@param dictDat: 配置数据 { "Param1":"", "Param2":"", "Param3":"" }
		@type  dictDat: python dict
		"""
		HoldEffect.init( self, dictDat, skill )
		actionInfos = dictDat["Param1"].split("|")
		for i in actionInfos:
			actionInfo = i.split(":")
			if len( actionInfo ) > 1:
				self._actionID.append(actionInfo[0])
				self._actionBeginTime.append( actionInfo[1] )			
			else:
				self._actionID.append(actionInfo[0])
				self._actionBeginTime.append( "0.0")
				
		self._effectID = dictDat["Param2"]
		self._soundID = dictDat["Param3"]
	
	def onBegin( self, effectData, receiver ):
		"""
		效果开始
		"""
		HoldEffect.onBegin( self, effectData, receiver )
		if not effectData.getTempData("isReloadBuff", False):	#如果是重载buff，不重新加效果，因为在onClientActorCreate时会加
			receiver.setActionIndex()
			effectData.setTempData("actionIndex",receiver.actionIndex)
			receiver.allClients.HoldEffectPlayEffect( effectData.casterID, effectData.index, receiver.actionIndex, self._actionID, self._effectID, self._soundID, self._actionBeginTime )
	
	def onEnd( self, effectData, receiver ):
		"""
		效果结束
		"""
		receiver.allClients.HoldEffectEndEffect( effectData.index, effectData.getTempData("actionIndex", 0) )
		HoldEffect.onEnd( self, effectData, receiver )

	
	def onClientActorCreate( self, effectData, receiver, srcEntityID ):
		"""
		virtual method
		receiver在某客户端被创建出来
		"""
		HoldEffect.onClientActorCreate( self, effectData, receiver, srcEntityID )
		if srcEntityID == receiver.id:
			clientEntity = receiver.client
		else:
			playerEntity = KBEngine.entities.get( srcEntityID )
			if not playerEntity:
				return
			clientEntity = playerEntity.clientEntity( receiver.id )
		
		clientEntity.HoldEffectPlayEffect( effectData.casterID, effectData.index, effectData.getTempData("actionIndex", 0), self._actionID, self._effectID, self._soundID, self._actionBeginTime )		#CST-2568 恢复效果